rfcs/plugin-api/experiment-2.ts

```diff
@@ -0,0 +1,45 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable @typescript-eslint/no-empty-function */
+import { Config, EnvironmentFunctions } from "@inlang/core/src/config"
+
+export const myPlugin = createPlugin<{ pathPattern: string }>(({ pluginConfig, env }) => {
+	return {
+		id: "samuelstroschein.plugin-json",
+		defineConfig: (config) => {
+			if (pluginConfig.pathPattern === undefined) {
+				throw new Error("pathPattern is required")
+			}
+			config.readResources = readResources({ pluginConfig, env })
+			config.languages = getLanguages({ pluginConfig, env })
+		},
+	}
+})
+
+/**
+ * The function to configure a plugin.
+ *
+ * @example
+ *   plugins: [
+ * 	 	myPlugin({
+ * 	   	pathPattern: "hello",
+ * 	 	})
+ *   ]
+ */
+type PluginConfigFunction<PluginConfig> = (
+	pluginConfig: PluginConfig,
+) => (env: EnvironmentFunctions) => Plugin
+
+type Plugin = {
+	id: string
+	defineConfig(config: Partial<Config>): void
+}
+
+function createPlugin<PluginConfig>(
+	callback: (args: { pluginConfig: PluginConfig; env: EnvironmentFunctions }) => Plugin,
+): PluginConfigFunction<PluginConfig> {
+	return (pluginConfig) => (env) => callback({ pluginConfig, env })
```

Comment:
why does `env` and `config` get passed in two different steps? I would do it like in `experiment-3.ts`

Reply:
Experiment-3 requires us to increase the surface are of the plugin API manyfold. 

We would need to pass `env` to every function in `Plugin` instead of only once when creating a plugin. Passing the env once enables us an easier way forward to remove env in favor of better ergonomics in the future. We would only need to change `createPlugin` and not every function in `Plugin`.

Reply:
ah, got it!